### PR TITLE
[NFC][AArch64] Refactor AArch64LoopIdiomTransform in preparation for more idioms

### PR DESCRIPTION
**Pull request by david-arm**
In a future patch I intend to add support for other types of C loops that do memory comparisons with early exits, e.g.

```
  for (unsigned long i = start; i < end; i++) {
    if (p1[i] != p2[i])
      break;
  }
```

where we compare the loaded values prior to incrementing the induction variable `i`. This requires first refactoring the pass ready to support the new loop structures. I've created a new MemCompareIdiom class with support routines that can be reused by new code required to recognise these new loops.

I also modified MemCompareIdiom::generateMemCompare to ensure it is ready to support arbitrary induction variable types as well as supporting new comparison predicates when comparing the memory values.

### CODE REVIEW

File: llvm/lib/Target/AArch64/AArch64LoopIdiomTransform.cpp [hunk 1]
```diff
@@ -251,145 +375,90 @@ bool AArch64LoopIdiomTransform::recognizeByteCompare() {
...
-  // loop blocks, but not actually defined in the loop. This requires extra
-  // work setting up the byte.compare block, i.e. by introducing a select to
-  // choose the correct value.
-  // TODO: We could add support for this in future.
-  if (FoundBB == EndBB) {
-    for (PHINode &EndPN : EndBB->phis()) {
-      Value *WhileCondVal = EndPN.getIncomingValueForBlock(Header);
-      Value *WhileBodyVal = EndPN.getIncomingValueForBlock(WhileBB);
+  // If we match the pattern, IndPhi and Index will be replaced with the result
+  // of the mismatch. If any other instructions are used outside of the loop, we
```
Comment (sjoerdmeijer) [inline llvm/lib/Target/AArch64/AArch64LoopIdiomTransform.cpp:441]: I don't quite get this comment, the use of `match` and `mismatch`, I don't see what mismatch this exactly is referring to.

File: llvm/lib/Target/AArch64/AArch64LoopIdiomTransform.cpp [hunk 2]
```diff
@@ -187,6 +180,38 @@ AArch64LoopIdiomTransformPass::run(Loop &L, LoopAnalysisManager &AM,
 //
 //===----------------------------------------------------------------------===//
 
+struct MemCompareIdiom {
+private:
```
Comment (mshockwave) [inline llvm/lib/Target/AArch64/AArch64LoopIdiomTransform.cpp:184]: Could we use class here instead?

File: llvm/lib/Target/AArch64/AArch64LoopIdiomTransform.cpp [hunk 3]
```diff
@@ -251,145 +375,90 @@ bool AArch64LoopIdiomTransform::recognizeByteCompare() {
...
-  // Where the incoming values for %final_ptr are unique and from each of the
-  // loop blocks, but not actually defined in the loop. This requires extra
-  // work setting up the byte.compare block, i.e. by introducing a select to
-  // choose the correct value.
-  // TODO: We could add support for this in future.
-  if (FoundBB == EndBB) {
-    for (PHINode &EndPN : EndBB->phis()) {
-      Value *WhileCondVal = EndPN.getIncomingValueForBlock(Header);
-      Value *WhileBodyVal = EndPN.getIncomingValueForBlock(WhileBB);
+  // If we match the pattern, IndPhi and Index will be replaced with the result
```
Comment (mshockwave) [inline llvm/lib/Target/AArch64/AArch64LoopIdiomTransform.cpp:440]: "...IndPhi and Index will be replaced with the result of the mismatch..." IIUC, this comment is outdated as this transformation no longer replaces `IndPhi` with mismatch result. Specifically, I think an older version of this patch RAUW on `IndPhi` & `Index` around line 815, where we only do that for `Index` now.

File: llvm/lib/Target/AArch64/AArch64LoopIdiomTransform.cpp (hunk 2 above)
Comment (david-arm) [reply llvm/lib/Target/AArch64/AArch64LoopIdiomTransform.cpp:184]: We could, but I saw lots of other places in LLVM where we use struct for smaller tasks like this so I thought maybe I should do the same. To be honest, I'm not entirely sure what the difference is. I'm happy to change it to class if that brings benefits!

File: llvm/lib/Target/AArch64/AArch64LoopIdiomTransform.cpp [hunk 4]
```diff
@@ -202,32 +227,135 @@ bool AArch64LoopIdiomTransform::run(Loop *L) {
                     << CurLoop->getHeader()->getParent()->getName()
                     << "] Loop %" << CurLoop->getHeader()->getName() << "\n");
 
-  return recognizeByteCompare();
+  MemCompareIdiom BCI(TTI, DT, LI, L);
+  if (BCI.recognize()) {
```
[the remark this answers is not in the document]
Comment (david-arm) [reply llvm/lib/Target/AArch64/AArch64LoopIdiomTransform.cpp:231]: I wanted to leave the door open for recognising more types than just bytes. Also, given the class itself is called `MemCompareIdiom` I was worried that `MemCompareIdiom::recognizeByteCompare` was a bit wordy, but I'm happy to  change it to `recognizeMemCompare` instead of just `recognize` if that helps?

File: llvm/lib/Target/AArch64/AArch64LoopIdiomTransform.cpp [hunk 5]
```diff
@@ -202,32 +227,135 @@ bool AArch64LoopIdiomTransform::run(Loop *L) {
...
+  if (std::distance(BB2.begin(), BB2.end()) > Block2Size)
+    return false;
+
+  return true;
+}
+
+bool MemCompareIdiom::checkIterationCondition(BasicBlock *CheckItBB,
+                                              BasicBlock *&LoopBB) {
+  ICmpInst::Predicate Pred;
+  if (!match(CheckItBB->getTerminator(),
```
[the remark this answers is not in the document]
Comment (david-arm) [reply llvm/lib/Target/AArch64/AArch64LoopIdiomTransform.cpp:296]: I think I might have looked at this before and it requires loops to have a single latch. The icmp in this case occurs in the first block, not the last block. I'll double check, but I think there was a reason I couldn't use that.

File: llvm/lib/Target/AArch64/AArch64LoopIdiomTransform.cpp [hunk 6]
```diff
@@ -251,145 +375,90 @@ bool AArch64LoopIdiomTransform::recognizeByteCompare() {
...
-      !LoadAI->getType()->isIntegerTy(8) ||
-      !LoadBI->getType()->isIntegerTy(8) || PtrA == PtrB)
+  // In AArch64LoopIdiomTransform::run we have already checked that the loop
+  // has a preheader so we can assume it's in a canonical form.
+  if (CurLoop->getNumBackEdges() != 1 || CurLoop->getNumBlocks() != 2)
     return false;
 
-  // Check that the index to the GEPs is the index we found earlier
-  if (GEPA->getNumIndices() > 1 || GEPB->getNumIndices() > 1)
+  // We only expect one PHI node for the index.
```
[the remark this answers is not in the document]
Comment (david-arm) [reply llvm/lib/Target/AArch64/AArch64LoopIdiomTransform.cpp:420]: Possibly? Although `getInductionVariable` calls `getLatchCmpInst` so I think we might have the same issue as mentioned above.

File: llvm/lib/Target/AArch64/AArch64LoopIdiomTransform.cpp (hunk 3 above)
Comment (david-arm) [reply llvm/lib/Target/AArch64/AArch64LoopIdiomTransform.cpp:440]: Yes I think you're right. I'll update the comment!

File: llvm/lib/Target/AArch64/AArch64LoopIdiomTransform.cpp (hunk 1 above)
Comment (david-arm) [reply llvm/lib/Target/AArch64/AArch64LoopIdiomTransform.cpp:441]: I think it refers to initial versions of the patch - I'll fix it!

File: llvm/lib/Target/AArch64/AArch64LoopIdiomTransform.cpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (david-arm) [reply llvm/lib/Target/AArch64/AArch64LoopIdiomTransform.cpp:296]: So this function does return something, but if you don't mind I've not used it here because I'd still have to do the work to extract the blocks in the branch, check the icmp predicate, Index, MaxLen, etc. I think the code is more compact right now, whereas using the latch function would require something like this:

```
  ICmpInst IC = CurLoop->getLatchCmpInst();
  if (!IC ||
      !match(CheckItBB->getTerminator(), m_Br(m_Specific(IC), m_BasicBlock(EndBB), m_BasicBlock(LoopBB))) ||
      IC->getOperand(0) != Index ||
      IC->getPredicate() != ICmpInst::Predicate::ICMP_EQ ||
      !CurLoop->contains(LoopBB))
     return false;

  MaxLen = IC->getOperand(1);
  return false;
```

File: llvm/lib/Target/AArch64/AArch64LoopIdiomTransform.cpp (hunk 6 above)
[the remark this answers is not in the document]
Comment (david-arm) [reply llvm/lib/Target/AArch64/AArch64LoopIdiomTransform.cpp:419]: So I think this is possible, but again if you don't mind I've kept the original code. The reason is that `getInductionVariable` requires initialising the ScalarEvolution class, doing work to create SCEVAddRecAddr and InductionDescriptor objects for nodes that we're going to throw away and ignore. It's also far more powerful than we need, since it supports more induction variables than the narrow set we care about, i.e. it supports FP inductions, increments != 1, non-linear inductions, etc. It feels a shame to do all that work, and then afterwards restrict it to the very narrow set we actually care about for these tight loops.

**Comment by JeffHDF** [comment]
> In a future patch I intend to add support for other types of C loops that do memory comparisons with early exits, e.g.
> 
> ```
>   for (unsigned long i = start; i < end; i++) {
>     if (p1[i] != p2[i])
>       break;
>   }
> ```
> 
> where we compare the loaded values prior to incrementing the induction variable `i`. This requires first refactoring the pass ready to support the new loop structures. I've created a new MemCompareIdiom class with support routines that can be reused by new code required to recognise these new loops.
> 
> I also modified MemCompareIdiom::generateMemCompare to ensure it is ready to support arbitrary induction variable types as well as supporting new comparison predicates when comparing the memory values.

Hi, Where these more scenarios come from? How do you consider this question? As i known, there are some scenarios in sub-item 557.xz of SPEC2017 int. Will you consider to support these scenarios?

**Comment by david-arm** [comment]
> Hi, Where these more scenarios come from? How do you consider this question? As i known, there are some scenarios in sub-item 557.xz of SPEC2017 int. Will you consider to support these scenarios?

Yes, some of these new scenarios are in the xz benchmark in SPEC2017. There are loops recognised in blender too, although these are not in hot code. Certainly there are lots of cases in the LLVM test suite where we will transform loops too. To be honest, one of the main benefits of going wide and supporting more loops is to get better test coverage, although hopefully we will see increased performance in more applications as well.

**Comment by shaojingzhi** [comment]
Hi, I'm interested in this pattern too, and I wonder that when would you plan to finish this pattern match? Would you just recognize this pattern and transform it to the original while loop IR and do the following transform reusing the original code, or would you like to write something new for this pattern?

**Comment by david-arm** [comment]
> Hi, I'm interested in this pattern too, and I wonder that when would you plan to finish this pattern match? Would you just recognize this pattern and transform it to the original while loop IR and do the following transform reusing the original code, or would you like to write something new for this pattern?

I already have a second patch waiting to upstream once this patch lands. This patch just makes it a lot easier to recognise different forms of the same idiom and transform the loop using the same generation code. There are a surprisingly large number of variations of the same idiom out there in SPEC and other benchmarks!

File: llvm/lib/Target/AArch64/AArch64LoopIdiomTransform.cpp [hunk 7]
```diff
@@ -1,3 +1,4 @@
+
```
Comment (davemgreen) [inline llvm/lib/Target/AArch64/AArch64LoopIdiomTransform.cpp:1]: Remove this extra newline?

File: llvm/lib/Target/AArch64/AArch64LoopIdiomTransform.cpp [hunk 8]
```diff
@@ -592,7 +660,8 @@ Value *AArch64LoopIdiomTransform::expandFindMismatch(
   Value *SVERhsLoad = Builder.CreateMaskedLoad(SVELoadType, SVERhsGep, Align(1),
```
Comment (JeffHDF) [inline llvm/lib/Target/AArch64/AArch64LoopIdiomTransform.cpp:660]: Have you try to use normal load? And generate an epilogue loop after the SVE loop, I think there may be more performance impovement.
